### DISCUSSION TITLE
code-server on older web browsers.

### DISCUSSION THREAD

**Discussion by ghost** [coder/code-server#4416] [Q&A] [answered]
code-server doesn't seem to work on browsers older than a year (like Google Chrome 80 or earlier). When loading it with an older version you can login just fine but the workspace is completely white. What causes this?

EDIT: When checking the console it shows the error Failed to load resource: the server responded with a status of 404 (Not Found) workbench.js.

**Comment by lg-nineoneninethreenine** [comment, marked as answer]
Aha! Maybe the workspace JavaScript file is missing, so this happened. @jsjoeio can you locate a workspace js file in the codebase and share it here?

**Reply by jsjoeio** [reply]
Hmm...can you post a screenshot and the logs? Are you accessing via HTTPS?

**Reply by ghost** [reply]
I have tried HTTP and HTTPS.

![Screen Shot 2021-11-01 at 2 48 45 PM](https://user-images.githubusercontent.com/82516534/139725521-2028cfe5-8ec1-4f0e-b7b5-fcb7d2330235.png)
![Screen Shot 2021-11-01 at 2 45 52 PM](https://user-images.githubusercontent.com/82516534/139725524-aa71c4e7-a6f9-41f7-8b40-92c33c9f68df.png)

**Reply by jsjoeio** [reply]
What version of code-server are you using?

**Reply by jsjoeio** [reply]
Can you try clearing your cache as well?

**Reply by ghost** [reply]
> What version of code-server are you using?

I am using the latest version. I cleared the cache and tried this on multiple devices.

**Reply by ghost** [reply]
This seems to just be an issue with older browsers and not code-server.

With Chromium build 79 it tries to load: https://127.0.0.1/vs/code/browser/workbench/workbench.js
With more recent browsers it tries to load: https://127.0.0.1/static/b37ff28a0a582aee84a8f961755d0cb40a4081db/usr/lib/code-server/vendor/modules/code-oss-dev/out/vs/code/browser/workbench/workbench.js

**Reply by jsjoeio** [reply]
Nice debugging! I'm not sure it would try to load a different path on an older browser 🤔 Any ideas @TeffenEllis or @code-asher?

**Reply by code-asher** [reply]
This might be our use of optional chaining which is supported in Chrome 80+. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

If this files errors (due to the unsupported syntax) then it will never set the loader paths and so workbench.js will load from the wrong place.

This particular file is gone in the latest code so it might work now although I am not sure what kind of compatibility VS Code web has with older browsers so it might just be trading one problem for another.

**Reply by jsjoeio** [reply]
> This might be our use of optional chaining which is supported in Chrome 80+. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Nice sleuthing! I guess I thought we targeted `es5` but looks like we're targeting `es2020` [here](https://github.com/cdr/code-server/blob/main/tsconfig.json#L3). Is that why this would happen?

**Reply by code-asher** [reply]
Yup exactly.  We might still have this problem in the next release depending on the target VS Code is using.

**Reply by ghost** [reply]
Okay. That makes sense.

**Reply by whoyouhow** [reply]
> Yup exactly. We might still have this problem in the next release depending on the target VS Code is using.

Does that mean we cannot use code-server in old browsers like Chrome 80-?

**Reply by lg-nineoneninethreenine** [reply]
Perhaps. As JS and TS keep evolving, so should the browsers themselves. So, maybe that's why some webpages don't load on computers with old browsers, as the JS could be incompatible with the provided engine. Am I right?

**Reply by jsjoeio** [reply]
You can try but it's not guaranteed to work!

**Reply by whoyouhow** [reply]
I got a similar problem when I was trying to use code server on Chrome 72. It shows a blank page instead of 404, but the errors are identical. I think the errors are caused by ES6 grammar.
So is it possible to apply polyfills on the code server? Sorry for the silly question, but I'm not familiar with the structure of code server project.

**Reply by jsjoeio** [reply]
Hmm...maybe? It might be easier to fork it and then change the `tsconfig.json`. You may have to change it in `/lib/vscode` as well.
